### PR TITLE
sql: support NULLs in ARRAYs

### PR DESCRIPTION
**Pull request by justinj**

### CODE REVIEW

File: pkg/sql/sqlbase/table.go [hunk 1]
```diff
@@ -1579,27 +1615,59 @@ func MarshalColumnValue(col ColumnDescriptor, val parser.Datum) (roachpb.Value,
 		val.ResolvedType(), col.Type.SemanticType, col.Name)
 }
 
+func encodeArrayHeader(h arrayHeader, buf []byte) ([]byte, error) {
+	// The byte we append here is formatted as follows:
+	// * The high 4 bits encode the number of dimensions in the array.
```
Comment (knz) [inline pkg/sql/sqlbase/table.go]: I just realized something, sorry for being late about it: when designing encodings it's rather common+desirable to place the flags that are more likely to be zero in the more significant bits. The reason is that for when a human looks at a value it is good that the most common values also be the smaller in numeric value.

Hence my concrete suggestion to place the dimensions in the LSB and the null bit above that.

File: pkg/sql/sqlbase/table.go [hunk 2]
```diff
@@ -1579,27 +1615,59 @@ func MarshalColumnValue(col ColumnDescriptor, val parser.Datum) (roachpb.Value,
...
+	if err != nil {
+		return nil, err
+	}
+	header := arrayHeader{
+		hasNulls: d.HasNulls,
+		// TODO(justin): support multiple dimensions.
+		numDimensions: 1,
+		elementType:   elementType,
+		length:        uint64(d.Len()),
+		// We don't encode the NULL bitmap this way because we do it in lockstep with the
```
Comment (knz) [inline pkg/sql/sqlbase/table.go]: Clarify what "this way" means here. Also Go initializes fields to nil by default.

File: pkg/sql/sqlbase/table.go [hunk 3]
```diff
@@ -1579,27 +1615,59 @@ func MarshalColumnValue(col ColumnDescriptor, val parser.Datum) (roachpb.Value,
...
+	scratch, err = encodeArrayHeader(header, scratch)
 	if err != nil {
 		return nil, err
 	}
-	scratch = encoding.EncodeValueTag(scratch, encoding.NoColumnID, encType)
-	scratch = encoding.EncodeNonsortingUvarint(scratch, uint64(d.Len()))
-	for _, e := range d.Array {
+	nullBitmapStart := len(scratch)
+	if d.HasNulls {
+		for i := 0; i < len(d.Array)/8+1; i++ {
```
Comment (knz) [inline pkg/sql/sqlbase/table.go]: Explain the `+1`

File: pkg/sql/sqlbase/table.go [hunk 4]
```diff
@@ -1579,27 +1615,59 @@ func MarshalColumnValue(col ColumnDescriptor, val parser.Datum) (roachpb.Value,
...
+			scratch = append(scratch, 0)
+		}
+	}
+	for i, e := range d.Array {
 		var err error
-		scratch, err = encodeArrayElement(scratch, e)
-		if err != nil {
-			return nil, err
+		if d.HasNulls && e == parser.DNull {
+			scratch[nullBitmapStart+i/8] = scratch[nullBitmapStart+i/8] | (1 << uint(i%8))
```
Comment (knz) [inline pkg/sql/sqlbase/table.go]: capture this logic in a separate function (`setBit`) and document it

File: pkg/sql/sqlbase/table.go [hunk 5]
```diff
@@ -1669,6 +1733,10 @@ func encodeArrayElement(b []byte, d parser.Datum) ([]byte, error) {
 	case *parser.DOid:
 		return encoding.EncodeUntaggedIntValue(b, int64(t.DInt)), nil
 	}
+	if d == parser.DNull {
```
Comment (knz) [inline pkg/sql/sqlbase/table.go]: leave this condition out. The catch-all error below will reveal nulls just as well.

File: pkg/sql/sqlbase/table.go [hunk 6]
```diff
@@ -1212,37 +1212,73 @@ func DecodeTableValue(a *DatumAlloc, valType parser.Type, b []byte) (parser.Datu
...
+		Array:    make(parser.Datums, header.length),
 		ParamTyp: elementType,
 	}
 	var val parser.Datum
-	for i := uint64(0); i < n; i++ {
-		val, b, err = decodeUntaggedDatum(a, elementType, b)
-		if err != nil {
-			return nil, b, err
+	for i := uint64(0); i < header.length; i++ {
+		if header.hasNulls && header.isNull(i) {
```
Comment (knz) [inline pkg/sql/sqlbase/table.go]: move the `if hasNulls` condition inside `isNull` and simplify the call points?

File: pkg/sql/sqlbase/table.go [hunk 7]
```diff
@@ -1212,37 +1212,73 @@ func DecodeTableValue(a *DatumAlloc, valType parser.Type, b []byte) (parser.Datu
...
 	b = b[dataOffset:]
-	if len(b) < 2 {
-		return nil, b, errors.Errorf("buffer too small")
+	b, _, length, err := encoding.DecodeNonsortingUvarint(b)
+	if err != nil {
+		return arrayHeader{}, b, err
+	}
+	nullBitmap := []byte(nil)
+	if hasNulls {
+		nullBitmapNumBytes := length/8 + 1
```
Comment (knz) [inline pkg/sql/sqlbase/table.go]: capture this in a separate function (`makeBitVec`)

**Comment by justinj** [comment]
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/sql/sqlbase/table.go, line 1244 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17231#-KpzUKhplZzF9IYEDMup-r1-1244:-KpzvFt-BsmpQcJddIif:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec44f4046805a02b22d13c1face2fbb4a189ed80/pkg/sql/sqlbase/table.go#L1244)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

capture this in a separate function (`makeBitVec`) 
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/table.go, line 1273 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17231#-KpzUKhplZzF9IYEDMup-r1-1273:-KpzvLLaNeoxEJuj_Z9O:b-60f218) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec44f4046805a02b22d13c1face2fbb4a189ed80/pkg/sql/sqlbase/table.go#L1273)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

move the `if hasNulls` condition inside `isNull` and simplify the call points?
</blockquote></details>

:thumbsup: way better

---

*[pkg/sql/sqlbase/table.go, line 1620 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17231#-KpzUKhplZzF9IYEDMup-r1-1620:-Kpzns7smS5BzSEQjRXH:bwgw6z0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec44f4046805a02b22d13c1face2fbb4a189ed80/pkg/sql/sqlbase/table.go#L1620)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I just realized something, sorry for being late about it: when designing encodings it's rather common+desirable to place the flags that are more likely to be zero in the more significant bits. The reason is that for when a human looks at a value it is good that the most common values also be the smaller in numeric value.

Hence my concrete suggestion to place the dimensions in the LSB and the null bit above that.
</blockquote></details>

That is a very cool point I had never heard! Will do.

---

*[pkg/sql/sqlbase/table.go, line 1648 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17231#-KpzUKhplZzF9IYEDMup-r1-1648:-Kpzv_nFvKqjHsPc09Im:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec44f4046805a02b22d13c1face2fbb4a189ed80/pkg/sql/sqlbase/table.go#L1648)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Clarify what "this way" means here. Also Go initializes fields to nil by default.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/table.go, line 1658 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17231#-KpzUKhplZzF9IYEDMup-r1-1658:-KpzvaOJE7jRv3yuQnXD:bfa78mo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec44f4046805a02b22d13c1face2fbb4a189ed80/pkg/sql/sqlbase/table.go#L1658)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Explain the `+1`
</blockquote></details>

Ergh, you're right, should be `(len+7)/8` :/

---

*[pkg/sql/sqlbase/table.go, line 1665 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17231#-KpzUKhplZzF9IYEDMup-r1-1665:-Kq--Y4BBGS3L07TSZCo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec44f4046805a02b22d13c1face2fbb4a189ed80/pkg/sql/sqlbase/table.go#L1665)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

capture this logic in a separate function (`setBit`) and document it
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/table.go, line 1736 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17231#-KpzUKhplZzF9IYEDMup-r1-1736:-Kq-0dy5IWy843XbWNXY:bhos1f5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec44f4046805a02b22d13c1face2fbb4a189ed80/pkg/sql/sqlbase/table.go#L1736)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

leave this condition out. The catch-all error below will reveal nulls just as well.
</blockquote></details>

Yeah, fair. I debated panicking here because it shouldn't happen, but that seemed like overkill.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17231)*

**Comment by knz** [comment]
LGTM modulo fixing the test.

---

Reviewed 1 of 4 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/table_test.go, line 336 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17231#-Kq28WwNO1iC883H-R5A:-Kq28WwOkBA-0Y0iAKgL:b-hqynp5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d8af86bb39d7bd5c55f2b2e6853986cdcc0696a/pkg/sql/sqlbase/table_test.go#L336)):*
> ```Go
> 			enc, err := encodeArray(&test.datum, nil)
> 			if err != nil {
> 				t.Error(err)
> ```

if `err != nil` then the `Equal` below shouldn't run.

I think `Fatal` was useful here. Also FYI `Fatal` in a sub-test only aborts that sub-test.

---

*[pkg/sql/sqlbase/table_test.go, line 349 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17231#-Kq28Y0Khm1lee4nFe_Q:-Kq28Y0Khm1lee4nFe_R:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d8af86bb39d7bd5c55f2b2e6853986cdcc0696a/pkg/sql/sqlbase/table_test.go#L349)):*
> ```Go
> 			d, _, err := decodeArray(&DatumAlloc{}, test.datum.ParamTyp, enc)
> 			if err != nil {
> 				t.Error(err)
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17231#-:-Kq28ju9YqIFrjhwVpCg:b58swxp)*

**Comment by justinj** [comment]
Thanks for the review!

**Pull request merged**